# "Sidetable, el beffo de pandas"
> "Sidetable es un paquete que te ofrece unos métodos extras para tus DataFrames de pandas."

- cover: assets/blogposts/019.png
- navigation: True
- date: 2020-07-01 10:00:00
- tags: [blog, pandas, python]
- class: post-template
- subclass: 'post tag-blog'
- author: sergio

`Sidetable` es una biblioteca de python que agrega métodos a los DataFrames de pandas para facilitar la exploración de datos. Según Chris Moffit (el creador de `sidetable`), en el repositorio de GitHub del proyecto, `sidetables` es como una versión más chida de `.value_counts()` y `.crosstab()` juntos para facilitar la creación de tablas resumen (? _summary tables_). [Repositorio de `sidetable` en GitHub](https://github.com/chris1610/sidetable#usage)

Este tutorial va a ser una rápida introducción a `sidetable` para demostrar sus habilidades. Estaremos utilizando el conjunto de datos `palmerpenguins` que tiene datos sobre 💖PINGÜINOS💖 que es una gran alternativa al dataset `iris` que es comúnmente usado en estos tutoriales por que 1) los datos vienen de un estadístico racista que los publicó en una publicación sobre eugenesia🤮

<blockquote class="twitter-tweet"><p lang="en" dir="ltr">me: using the iris dataset is fine. not everything Fisher did was about eugenics!<br><br>me: *looks up the iris dataset citation* 😮 <a href="https://t.co/GWymD6IG7I">pic.twitter.com/GWymD6IG7I</a></p>&mdash; 🔥Kareem Carr🔥 (@kareem_carr) <a href="https://twitter.com/kareem_carr/status/1271096239103369224?ref_src=twsrc%5Etfw">June 11, 2020</a></blockquote> <script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script>

Para más información de los datos visita su página oficial [allisonhorst.github.io/palmerpenguis/](https://allisonhorst.github.io/palmerpenguins/)

Para nuestra suerte la biblioteca `seaborn` y su complemento `seaborn-data` nos ayudan a obtener los datos rápidamente.

# Prepareishon

Obvis microbis tines que tener instalado `pandas` y `sidetable`. en este tutorial vamos a usar `seaborn` para acceder los datos entonces también necesitarás instalarlo.

```shell
pip install -U pandas sidetable seaborn
```

la `U` es de `--upgrade`, significa que actualiza las bibliotecas si ya las tienes instaladas.

In [8]:
import seaborn as sns
import pandas as pd
import sidetable

print(f"Versión de seaborn: {sns.__version__}")
print(f"Versión de pandas: {pd.__version__}")
print(f"Versión de sidetable: {sidetable.__version__}")

Versión de seaborn: 0.9.0
Versión de pandas: 1.0.5
Versión de sidetable: 0.5.0


In [9]:
datos = sns.load_dataset('penguins')

datos.head()

species     island  culmen_length_mm  culmen_depth_mm  flipper_length_mm  \
0  Adelie  Torgersen              39.1             18.7              181.0   
1  Adelie  Torgersen              39.5             17.4              186.0   
2  Adelie  Torgersen              40.3             18.0              195.0   
3  Adelie  Torgersen               NaN              NaN                NaN   
4  Adelie  Torgersen              36.7             19.3              193.0   

   body_mass_g     sex  
0       3750.0    MALE  
1       3800.0  FEMALE  
2       3250.0  FEMALE  
3          NaN     NaN  
4       3450.0  FEMALE

# Exploreishon

Dos métodos para explorar un DataFrame rápidamente son `.info()` y `.describe()` los cuales te muestran lo siguiente

In [10]:
datos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   species            344 non-null    object 
 1   island             344 non-null    object 
 2   culmen_length_mm   342 non-null    float64
 3   culmen_depth_mm    342 non-null    float64
 4   flipper_length_mm  342 non-null    float64
 5   body_mass_g        342 non-null    float64
 6   sex                333 non-null    object 
dtypes: float64(4), object(3)
memory usage: 18.9+ KB


In [11]:
datos.describe()

culmen_length_mm  culmen_depth_mm  flipper_length_mm  body_mass_g
count        342.000000       342.000000         342.000000   342.000000
mean          43.921930        17.151170         200.915205  4201.754386
std            5.459584         1.974793          14.061714   801.954536
min           32.100000        13.100000         172.000000  2700.000000
25%           39.225000        15.600000         190.000000  3550.000000
50%           44.450000        17.300000         197.000000  4050.000000
75%           48.500000        18.700000         213.000000  4750.000000
max           59.600000        21.500000         231.000000  6300.000000

`.info()` te muestra una tabla con las columnas de tu DataFrame, la cuenta de valores _no nulos_ y el _dtype_ o _data type_ de los datos de la columna. Mientras que `describe()` te muestra estadísticas de tus columnas numéricas. La cuenta, el promedio, la deviación estandar, el mínimo y máximo y los percentiles 25, 50 (la mediana) y 75.

## Valores nulos

Lo que yo hacía antes era utilizar el método `.isna()` o `.isnull()` en el DataFrame _encadenado_ a `.sum()` para ver cuantos valores nulos tenía cada columna

In [12]:
datos.isna().sum()

species               0
island                0
culmen_length_mm      2
culmen_depth_mm       2
flipper_length_mm     2
body_mass_g           2
sex                  11
dtype: int64

Y luego si quería saber que porcentage es del total era hacer algo así

In [15]:
datos.isna().sum() / len(datos)

species              0.000000
island               0.000000
culmen_length_mm     0.005814
culmen_depth_mm      0.005814
flipper_length_mm    0.005814
body_mass_g          0.005814
sex                  0.031977
dtype: float64

ya que `len(datos)` te da el "largo" del DataFrame, osea el número de filas.

`sidetable` simplifica esto de la siguiente manera

In [16]:
datos.stb.missing()

Missing  Total   Percent
sex                     11    344  0.031977
culmen_length_mm         2    344  0.005814
culmen_depth_mm          2    344  0.005814
flipper_length_mm        2    344  0.005814
body_mass_g              2    344  0.005814
species                  0    344  0.000000
island                   0    344  0.000000

y hasta lo enchula 

In [17]:
datos.stb.missing(style = True)

La mágia de `sidetable` es que agrega un _accessor_ a los DataFrames de pandas. Al hacer `import sidetable` estás habilitando el _accessor_ `.stb`. Así es como se integra a tu flujo de trabajo.

## Frecuencias

El siguiente método incluido en `sidetable` es `.freq()` que es como un `.value_counts()` súper poderoso.

In [21]:
datos['island'].value_counts()

Biscoe       168
Dream        124
Torgersen     52
Name: island, dtype: int64

In [22]:
datos['island'].value_counts(normalize = True)

Biscoe       0.488372
Dream        0.360465
Torgersen    0.151163
Name: island, dtype: float64

In [28]:
datos.stb.freq(['island'])

island  Count   Percent  Cumulative Count  Cumulative Percent
0     Biscoe    168  0.488372               168            0.488372
1      Dream    124  0.360465               292            0.848837
2  Torgersen     52  0.151163               344            1.000000

Este método también tiene el parametro `style` para limpiar la columna de porcentages.

In [29]:
datos.stb.freq(['island'], style = True)

Algo que me encantó de esta biblioteca es que crea también estos porcentages y cuentas cumulativas. En este caso no agrega mucho contexto pero cuando trabajo con datos censales se me hace muy útil.

Habrás notado que le pasamos una lista de columnas (aunque haya sido una lista de 1. De hecho, el parámetro de la función `.freq()` se llama `cols` en plural. Veamos que sucede si le pasamos más de una columna.

In [31]:
datos.stb.freq(['island', 'species', 'sex'], style = True)

para lograr algo similar puro pandas tendría que hacer algo como

In [39]:
datos.groupby(['island', 'species', 'sex'])['culmen_length_mm'].count().to_frame().reset_index()

island    species     sex  culmen_length_mm
0     Biscoe     Adelie  FEMALE                22
1     Biscoe     Adelie    MALE                22
2     Biscoe     Gentoo  FEMALE                58
3     Biscoe     Gentoo    MALE                61
4      Dream     Adelie  FEMALE                27
5      Dream     Adelie    MALE                28
6      Dream  Chinstrap  FEMALE                34
7      Dream  Chinstrap    MALE                34
8  Torgersen     Adelie  FEMALE                24
9  Torgersen     Adelie    MALE                23

guardar este DataFrame con un nombre nuevo y crearle nuevas columnas. Con `sidetable` logro esto con un sólo comando.

Esa es la mágia de `sidetable` en mi opinión. No es que este ofreciendote funcionalidades que no existían en pandas. Lo que hace es hacerlas más accesibles 💖

Este método `.freq()` también tiene el parámetro `thresh` que sirve para agrupar valores hasta cierto valor cumulativo. 

In [41]:
datos.stb.freq(['island', 'species'], style = True)

In [42]:
datos.stb.freq(['island', 'species'], style = True, thresh = .6)

Lo que `sidetable` esta haciendo es mostrar tus valores hasta que llegan a tu `thresh` o límite. En este caso, muestra todos los valores que van sumando hasta llegar a un valor cumulativo de 60% y los demás valores se agrupan bajo la etiqueta `Others` (la cual también puedes modificar).

In [43]:
datos.stb.freq(['island', 'species'], style = True, thresh = .6, other_label = "baia baia tacubaya")

## Subtotal

El tercer método que `sidetable` ofrece es `.subtotal()` que te permite, como su nombre lo indica, crear subtotales de tus datos agrupados.

In [45]:
datos.stb.subtotal()

species     island  culmen_length_mm  culmen_depth_mm  \
0            Adelie  Torgersen              39.1             18.7   
1            Adelie  Torgersen              39.5             17.4   
2            Adelie  Torgersen              40.3             18.0   
3            Adelie  Torgersen               NaN              NaN   
4            Adelie  Torgersen              36.7             19.3   
...             ...        ...               ...              ...   
340          Gentoo     Biscoe              46.8             14.3   
341          Gentoo     Biscoe              50.4             15.7   
342          Gentoo     Biscoe              45.2             14.8   
343          Gentoo     Biscoe              49.9             16.1   
Grand Total     NaN        NaN           15021.3           5865.7   

             flipper_length_mm  body_mass_g     sex  
0                        181.0       3750.0    MALE  
1                        186.0       3800.0  FEMALE  
2                        195.0       3250.0  FEMALE  
3                          NaN          NaN     NaN  
4                        193.0       3450.0  FEMALE  
...                        ...          ...     ...  
340                      215.0       4850.0  FEMALE  
341                      222.0       5750.0    MALE  
342                      212.0       5200.0  FEMALE  
343                      213.0       5400.0    MALE  
Grand Total            68713.0    1437000.0     NaN  

[345 rows x 7 columns]

Puedes ver como agrega la fila **Grand total** a tu DataFrame. No es como que digas hijole que bruto que útil agregar un total a esta tabla pero lo más interesante aparece cuando aplicas `.subtotal()` a un objeto `.groupby()`

In [46]:
datos.groupby(['species', 'sex'])[['island']].count()

island
species   sex           
Adelie    FEMALE      73
          MALE        73
Chinstrap FEMALE      34
          MALE        34
Gentoo    FEMALE      58
          MALE        61

In [49]:
datos.groupby(['species', 'sex'])[['island']].count().stb.subtotal()

island
species     sex                         
Adelie      FEMALE                    73
            MALE                      73
            Adelie - subtotal        146
Chinstrap   FEMALE                    34
            MALE                      34
            Chinstrap - subtotal      68
Gentoo      FEMALE                    58
            MALE                      61
            Gentoo - subtotal        119
Grand Total                          333

Para hacer algo así en pandas tendrías que hacer algo como

In [58]:
datos.pivot_table(index=['species', 'sex',], values="culmen_length_mm", aggfunc='count', margins = True)

culmen_length_mm
species   sex                     
Adelie    FEMALE                73
          MALE                  73
Chinstrap FEMALE                34
          MALE                  34
Gentoo    FEMALE                58
          MALE                  61
All                            333

***

Esas son las 3 funcionalidades que `sidetable` agrega a pandas: `.missing()`, `.freq()` y `.subtotal()`. 

Personalmente, voy a utilizar `sidetable` en todos los cursos y talleres que imparta de ahora en adelante. Ofrece el poder de pandas pero más accesible. 


***

## Bonus

Aunque me encanten los pingüinos la verdad es que yo no uso ese tipo de datos en mi día a día así que aquí hay un ejemplo de una tabla que produciría en el trabajo. 

Para este ejercicio voy a utilizar `pypums` una biblioteca mia que te permite descargar datos del censo de estados unidos de manera automatizable con python.

In [59]:
import pandas as pd
import sidetable
from pypums import ACS

In [60]:
california = ACS(state = 'california', year = 2017)
data = california.as_dataframe()
data.head()

RT       SERIALNO  DIVISION  SPORDER  PUMA  REGION  ST   ADJINC  PWGTP  \
0  P  2017000000014         9        1  6703       4   6  1011189     59   
1  P  2017000000014         9        2  6703       4   6  1011189     57   
2  P  2017000000014         9        3  6703       4   6  1011189     54   
3  P  2017000000017         9        1   110       4   6  1011189     75   
4  P  2017000000017         9        2   110       4   6  1011189     71   

   AGEP  ...  PWGTP71  PWGTP72  PWGTP73  PWGTP74  PWGTP75  PWGTP76  PWGTP77  \
0    45  ...       64       63       55       17      100       65       62   
1    39  ...       51       54       59       19       95       53       56   
2    15  ...       50       62       52       14       99       53       60   
3    34  ...      140       73      127       67       21       73       27   
4    33  ...      116       75      127       73       19       79       25   

   PWGTP78  PWGTP79  PWGTP80  
0       68       17      101  
1       62       18       97  
2       60       16       90  
3      117       65      113  
4      110       72      123  

[5 rows x 286 columns]

Usando `sidetable` puedo encontrar estadísticas sobre la población de California rápidamente. Algo que me gusta mucho de `sidetable` es que puedo agregar el parámetro `value` para pasarle una columna que va a utilizar para la agregación de mis datos. En este caso, uso `PWGTP` (_person weight_) o factor de expansión. Esto es "lo que vale" cada observación en mis datos ya que estos datos son una muestra de la población total de estados unidos (es la American Community Survey que encuesta más o menos 1% de la población anualmente).

In [61]:
data.stb.freq(['SEX'], value = 'PWGTP', style = True)

Algo que yo he hecho muchas veces en mi trabajo anterior era investigar cuantas personas en California de X edad son inmigrantes, por ejemplo. Este número serviría de denominador en otras estadísticas. 

Ya que tenemos los datos, agrupo la columna `AGEP` que tiene los valores de las edades.

In [62]:
# 0 - 17
mask_17 = data['AGEP'] < 18
data.loc[mask_17, 'AGE_BIN'] = '0 - 17'

# 18 - 25
mask_25 = (data['AGEP'] >= 18) & (data['AGEP'] <= 25)
data.loc[mask_25, 'AGE_BIN'] = '18 - 25'

# 26 - 35
mask_35 = (data['AGEP'] >= 26) & (data['AGEP'] <= 35)
data.loc[mask_35, 'AGE_BIN'] = '26 - 35'

# 36 - 45
mask_45 = (data['AGEP'] >= 36) & (data['AGEP'] <= 45)
data.loc[mask_45, 'AGE_BIN'] = '36 - 45'

# 46 - 55
mask_55 = (data['AGEP'] >= 46) & (data['AGEP'] <= 55)
data.loc[mask_55, 'AGE_BIN'] = '46 - 55'

# 56 - 65
mask_65 = (data['AGEP'] >= 56) & (data['AGEP'] <= 65)
data.loc[mask_65, 'AGE_BIN'] = '56 - 65'

# 65+
mask_65plus = (data['AGEP'] >= 66)
data.loc[mask_65plus, 'AGE_BIN'] = '65+'

In [63]:
age_bins = [
    '0 - 17',
    '18 - 25',
    '26 - 35',
    '36 - 45',
    '46 - 55',
    '56 - 65',
    '65+',
]
data['AGE_BIN'] = pd.Categorical(data['AGE_BIN'], categories = age_bins, ordered = True)

Y ya podemos ver por ejemplo, cuantos hombres y mujeres (abajo con el binarismo 👎) hay en cada grupo.

In [64]:
data.stb.freq(['SEX', 'AGE_BIN'], value = 'PWGTP', style = True)

Para limpiar los datos un poco (para presentarlos en alguna junta o algo) puedo cambiar los valores (que obtuve del diccionario de datos: https://www2.census.gov/programs-surveys/acs/tech_docs/pums/data_dict/PUMS_Data_Dictionary_2018.pdf?#)

In [65]:
data_dict_sex = {
    1: 'Male',
    2: 'Female',
}

data_dict_cit = {
    1: "Born in the U.S.",
    2: "Born in Puerto Rico, Guam, the U.S. Virgin Islands, or the Northern Marianas",
    3: "Born abroad of American parent(s)",
    4: "U.S. citizen by naturalization",
    5: "Not a citizen of the U.S.",
}

In [66]:
data['SEX'] = data['SEX'].map(data_dict_sex)
data['CIT'] = data['CIT'].map(data_dict_cit)

In [67]:
data['CIT'] = pd.Categorical(data['CIT'], categories = data_dict_cit.values(), ordered = True)

¡Y listo!

In [68]:
data.stb.freq(['CIT', 'AGE_BIN', 'SEX'], value = 'PWGTP', style = True, sort_cols=True)

O si quiero solamente aquellas personas que no nacieron en estados unidos

In [69]:
mask_noncitizen = data['CIT'] == 'Not a citizen of the U.S.'
data[mask_noncitizen].stb.freq(['CIT', 'AGE_BIN', 'SEX'], value = 'PWGTP', style = True, sort_cols=True)

¡Una visualización rapidín!

In [76]:
import altair as alt

# nota que le no estoy usando style = True
vis_data = data[mask_noncitizen].stb.freq(['CIT', 'AGE_BIN', 'SEX'], value = 'PWGTP', sort_cols=True)

alt.Chart(vis_data).mark_bar().encode(
    x='PWGTP:Q',
    y='SEX:O',
    color='SEX:N',
    row='AGE_BIN:O'
).properties(
    title = "No. de personas no nacidas en EEUU en California (ACS 2017)"
)

alt.Chart(...)

En conclusión, `sidetable` es una herramienta sencilla, fácil de usar y muy útil.

**Créditos:**
Aprendí de `sidetable` en el podcast _python bytes_ en el [episodio #186](https://pythonbytes.fm/episodes/show/186/the-treebeard-will-guard-your-notebook) pero mientras escribía esta nota encontré este blog en twitter que también esta explorando `sidetable` pero en inglés: https://beta.deepnote.com/article/sidetable-pandas-methods-you-didnt-know-you-needed así que gracias a ambos recursos 🤓

***

¿Qué te pareció la nota? [Mandanos un tuit a @tacosdedatos](https://twitter.com/share?text=Obvio+que+estuvo+super+el+blog+%40tacosdedatos+%F0%9F%8C%AE) o envianos un correo a [✉️ sugerencias@tacosdedatos.com](mailto:sugerencias@tacosdedatos.com?subject=Sugerencia&body=Hola-holaaa). Y recuerda que puedes subscribirte a nuestro boletín semanal aquí debajo.